hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/catalog/HoodieHiveCatalog.java

```diff
@@ -0,0 +1,894 @@
+  public void renameTable(ObjectPath tablePath, String newTableName, boolean ignoreIfNotExists)
+      throws TableNotExistException, TableAlreadyExistException, CatalogException {
+    checkNotNull(tablePath, "tablePath cannot be null");
+    checkArgument(
+        !isNullOrWhitespaceOnly(newTableName), "newTableName cannot be null or empty");
+
+    try {
+      // alter_table() doesn't throw a clear exception when target table doesn't exist.
+      // Thus, check the table existence explicitly
+      if (tableExists(tablePath)) {
+        ObjectPath newPath = new ObjectPath(tablePath.getDatabaseName(), newTableName);
+        // alter_table() doesn't throw a clear exception when new table already exists.
+        // Thus, check the table existence explicitly
+        if (tableExists(newPath)) {
+          throw new TableAlreadyExistException(getName(), newPath);
+        } else {
+          Table hiveTable = getHiveTable(tablePath);
+
+          //update hoodie
+          StorageDescriptor sd = hiveTable.getSd();
+          String location = sd.getLocation();
+          HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder().setBasePath(location).setConf(hiveConf).build();
+          //Init table with new name
+          HoodieTableMetaClient.withPropertyBuilder().fromProperties(metaClient.getTableConfig().getProps())
+              .setTableName(newTableName)
+              .initTable(hiveConf, location);
+
+          hiveTable.setTableName(newTableName);
+          client.alter_table(
+              tablePath.getDatabaseName(), tablePath.getObjectName(), hiveTable);
+        }
+      } else if (!ignoreIfNotExists) {
+        throw new TableNotExistException(getName(), tablePath);
+      }
+    } catch (Exception e) {
+      throw new HoodieCatalogException(
+          String.format("Failed to rename table %s", tablePath.getFullName()), e);
+    }
+  }
+
+  @Override
+  public void alterTable(
+      ObjectPath tablePath, CatalogBaseTable newCatalogTable, boolean ignoreIfNotExists)
+      throws TableNotExistException, CatalogException {
+    checkNotNull(tablePath, "tablePath cannot be null");
+    checkNotNull(newCatalogTable, "newCatalogTable cannot be null");
+
+    if (!newCatalogTable.getOptions().getOrDefault(CONNECTOR.key(), "").equalsIgnoreCase("hudi")) {
+      throw new HoodieCatalogException(String.format("The %s is not hoodie table", tablePath.getObjectName()));
+    }
+    if (newCatalogTable instanceof CatalogView) {
+      throw new HoodieCatalogException("Hoodie catalog does not support to ALTER VIEW");
+    }
+
+    try {
+      Table hiveTable = getHiveTable(tablePath);
+    } catch (TableNotExistException e) {
+      if (!ignoreIfNotExists) {
+        throw e;
+      }
+      return;
+    }
+    try {
+      boolean isMorTable = newCatalogTable.getOptions().getOrDefault(FlinkOptions.TABLE_TYPE.key(),
+          FlinkOptions.TABLE_TYPE.defaultValue()).equalsIgnoreCase(FlinkOptions.TABLE_TYPE_MERGE_ON_READ);
```

Comment:
let‘s forbid  follow behavior
1) forbidden change hoodie table type . eg: mor -> cow /cow -> mor
2) if  'index.type' = 'BUCKET' is set already set for hoodie table.  forbidden change this property， since  bucket index is incompatible with other hoodie index

Reply:
Updated